### PR TITLE
Catch UCR non-nullable violations

### PR DESCRIPTION
**Pull request by emord**
@snopoke you asked for something like this in the original PR. I want to keep the amount of errors we catch small to start out with so that we can be sure there aren't other places where these errors are popping up.

### CODE REVIEW

**Comment by orangejenny** [comment]
> in the original PR

Which one was this?

**Comment by emord** [comment]
https://github.com/dimagi/commcare-hq/pull/23799#issuecomment-477471014

File: corehq/apps/userreports/sql/adapter.py [hunk 1]
```diff
@@ -260,8 +264,22 @@ class ErrorRaisingIndicatorSqlAdapter(IndicatorSqlAdapter):
...
         ex = translate_programming_error(exception)
-        if ex:
+        if ex is not None:
             raise ex
+
+        orig_exception = getattr(exception, 'orig')
+        if orig_exception and isinstance(orig_exception, psycopg2.IntegrityError):
+            if orig_exception.pgcode == psycopg2.errorcodes.NOT_NULL_VIOLATION:
+                from corehq.apps.userreports.models import InvalidUCRData
+                InvalidUCRData.objects.create(
```
Comment (orangejenny) [inline corehq/apps/userreports/sql/adapter.py]: Should this set `date_created`?

Comment (orangejenny) [reply corehq/apps/userreports/sql/adapter.py]: nm, answered my own question

File: corehq/apps/userreports/tests/test_save_errors.py [hunk 2]
```diff
@@ -68,6 +75,21 @@ def test_missing_column(self):
...
+        adapter.build_table()
+
+        doc = {
+            "_id": '123',
+            "domain": "domain",
+            "doc_type": "CommCareCase",
+            "name": None
+        }
+        adapter.best_effort_save(doc)
+        self.assertEqual(InvalidUCRData.objects.count(), 1)
```
Comment (orangejenny) [inline corehq/apps/userreports/tests/test_save_errors.py]: Asserting `validation_name` instead of or in addition to the count seems more robust.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/userreports/tests/test_save_errors.py (hunk 2 above)
Comment (emord) [reply corehq/apps/userreports/tests/test_save_errors.py]: good idea: https://github.com/dimagi/commcare-hq/pull/24023/commits/53b346ff3d816cc2e637dd19003ff06505023929

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/userreports/exceptions.py [hunk 3]
```diff
@@ -93,9 +95,7 @@ def translate_programming_error(exception):
     if isinstance(exception, ProgrammingError):
         orig = getattr(exception, 'orig')
         if orig:
-            error_code = getattr(orig, 'pgcode')
-            # http://www.postgresql.org/docs/9.4/static/errcodes-appendix.html
-            if error_code == '42P01':
+            if orig.pgcode == errorcodes.UNDEFINED_TABLE:
                 return TableNotFoundWarning(str(exception))
-            elif error_code == '42703':
+            elif orig.pgcode == errorcodes.UNDEFINED_COLUMN:
```
Comment (snopoke) [inline corehq/apps/userreports/exceptions.py]: :+1:

**Review by snopoke — COMMENTED** [review]
Is it not possible to validate this prior to saving? We already know which columns are not nullable from the the data source config so we could just check the row values base on that right?

**Comment by emord** [comment]
@snopoke I believe that we could do that and it would require:

1. Add method to `DataSourceConfiguration` to return non-nullable columns
2. in `get_all_values` in adapter (or datasource) parse dictionary to see if any of those columns are null
3. remove those rows from the database query and write to InvalidUCRData

That doesn't sound too hard to do and I could be persuaded to write that, but I do think this is simpler as it keeps those constraints in one place (the database) and in general those should be exceptional cases, not the norm.

Another edge case is if we have a base_item_expression that writes multiple rows. What happens if only one of those rows violates a non null expression? At the moment the entire document would not be written. Would we want to keep that behavior if catching them before the insert? We probably would want that unless we made changes to the model to define what iteration did not get saved

**Review by snopoke — APPROVED** [review]
Thanks, that all makes sense.

**Pull request merged**